File: internal/config/skips.go
package config

import (
	"fmt"
	"slices"
	"strings"
)

// The canonical skip keys name the pipeline work a single run may bypass. Each
// one covers exactly one Activity of a Step (ADR 0037): SkipLintFix and
// SkipCleanup close out Build, SkipVerify closes the whole Verify Step, and
// SkipCI and SkipMerge close two Activities of Ship. Build, commit and PR carry
// no key at all — a run that produces nothing has nothing to skip.
const (
	SkipLintFix = "lintfix"
	SkipCleanup = "cleanup"
	SkipVerify  = "verify"
	SkipCI      = "ci"
	SkipMerge   = "merge"
)

// skipKeys is the canonical set in the order the error message lists it.
var skipKeys = []string{SkipLintFix, SkipCleanup, SkipVerify, SkipCI, SkipMerge}

// SkipKeys returns the canonical skip keys, for callers that name them to a human.
func SkipKeys() []string { return append([]string(nil), skipKeys...) }

// CanonicalSkips reads an already-split skip set — a JSON array from the hub, a
// picker's answer — into the canonical keys it names. It rejects an unknown key
// naming the valid set, so a typo fails at the edge rather than silently running
// the work the operator meant to bypass. Blanks drop, repeats collapse, and the
// result keeps canonical order, so the same choice always records the same set;
// an empty set returns nil.
func CanonicalSkips(keys []string) ([]string, error) {
	named := map[string]bool{}
	for _, raw := range keys {
		key := strings.ToLower(strings.TrimSpace(raw))
		if key == "" {
			continue
		}
		if !ValidSkip(key) {
			return nil, fmt.Errorf("unknown key %q; valid keys are %s", raw, strings.Join(skipKeys, ", "))
		}
		named[key] = true
	}
	return canonicalSkips(named), nil
}

// ParseSkips reads one comma-separated --skip value through the same vocabulary
// the hub validates against, so the flag and the queue cannot disagree about
// what a key means.
func ParseSkips(v string) ([]string, error) {
	keys, err := CanonicalSkips(strings.Split(v, ","))
	if err != nil {
		return nil, fmt.Errorf("--skip: %w", err)
	}
	return keys, nil
}

// MergeSkips folds a second skip set into the one an earlier gesture produced —
// a repeated --skip occurrence, or a launch-time choice landing on the set an
// item was queued with — keeping the canonical order and dropping the repeats.
func MergeSkips(have, add []string) []string {
	named := map[string]bool{}
	for _, key := range append(append([]string(nil), have...), add...) {
		named[key] = true
	}
	return canonicalSkips(named)
}

// canonicalSkips renders a named set in canonical order.
func canonicalSkips(named map[string]bool) []string {
	var out []string
	for _, key := range skipKeys {
		if named[key] {
			out = append(out, key)
		}
	}
	return out
}

// ValidSkip reports whether key is one of the canonical skip keys.
func ValidSkip(key string) bool { return slices.Contains(skipKeys, key) }

File: web/src/lib/skips.ts
// The canonical skip keys, mirroring internal/config/skips.go: each one names
// exactly one Activity of a Step that a single run may bypass (ADR 0037). The
// hub refuses anything else, so this list is the whole vocabulary.
export const SKIP_LINTFIX = 'lintfix'
export const SKIP_CLEANUP = 'cleanup'
export const SKIP_VERIFY = 'verify'
export const SKIP_CI = 'ci'
export const SKIP_MERGE = 'merge'

// SKIP_KEYS is the canonical order the hub stores a set in, so a set built here
// round-trips through the queue unchanged.
export const SKIP_KEYS: string[] = [
  SKIP_LINTFIX,
  SKIP_CLEANUP,
  SKIP_VERIFY,
  SKIP_CI,
  SKIP_MERGE,
]

export interface PipelineActivity {
  label: string
  // skip is the canonical key that bypasses this Activity. An Activity without
  // one is locked on: a run that produces nothing has nothing to skip.
  skip?: string
  caption?: string
}

export interface PipelineStep {
  label: string
  activities: PipelineActivity[]
}

// PIPELINE_STEPS is what a run does, as the step picker lists it: the Activities
// grouped under the three Steps the stepper already shows. Verify is one row
// because its key closes the whole Step at once.
export const PIPELINE_STEPS: PipelineStep[] = [
  {
    label: 'Build',
    activities: [
      { label: 'Build', caption: 'writes the slice' },
      { label: 'Lint-fix', skip: SKIP_LINTFIX },
      { label: 'Cleanup', skip: SKIP_CLEANUP },
    ],
  },
  {
    label: 'Verify',
    activities: [
      {
        label: 'Verify',
        skip: SKIP_VERIFY,
        caption:
          'covers handoff brief, test gate, AI verification, repairs, browser verify',
      },
    ],
  },
  {
    label: 'Ship',
    activities: [
      { label: 'Commit', caption: 'commits the work' },
      { label: 'PR', caption: 'opens the pull request' },
      { label: 'CI wait', skip: SKIP_CI },
      { label: 'Auto-merge', skip: SKIP_MERGE },
    ],
  },
]

// canonicalSkips renders a named set in the canonical order and drops the
// repeats, so the same choice always sends the same set.
export function canonicalSkips(keys: Iterable<string>): string[] {
  const named = new Set(keys)
  return SKIP_KEYS.filter((key) => named.has(key))
}

// toggleSkip settles one Activity's checkbox: ticked means the run does the
// work, so it leaves the skip set; unticked adds it.
export function toggleSkip(
  skips: string[],
  key: string,
  enabled: boolean,
): string[] {
  const named = new Set(skips)
  if (enabled) named.delete(key)
  else named.add(key)
  return canonicalSkips(named)
}
